src/core/mesh/qgsmesh3daveraging.cpp

```diff
@@ -151,3 +248,513 @@ int QgsMeshSingleLevelAveragingMethod::verticalLevel() const
 {
   return mVerticalLevel;
 }
+
+bool QgsMeshSingleLevelAveragingMethod::hasValidInputs() const
+{
+  return mVerticalLevel >= 1;
+}
+
+void QgsMeshSingleLevelAveragingMethod::volumeRangeForFace(
+  double &startVerticalLevel,
+  double &endVerticalLevel,
+  const QVector<double> verticalLevels ) const
+{
+  if ( ( mVerticalLevel - 1 >= 0 ) && ( mVerticalLevel < verticalLevels.size() ) )
+  {
+    startVerticalLevel = verticalLevels[mVerticalLevel - 1];
+    endVerticalLevel = verticalLevels[ mVerticalLevel ];
+  }
+}
+
+QgsMeshMultiLevelsFromTopAveragingMethod::QgsMeshMultiLevelsFromTopAveragingMethod( int startLevel, int endLevel )
+  : QgsMesh3dAveragingMethod( QgsMesh3dAveragingMethod::MultiLevelsFromTopAveragingMethod )
+  , mStartVerticalLevel( startLevel )
+  , mEndVerticalLevel( endLevel )
+{
+  if ( mStartVerticalLevel > mEndVerticalLevel )
+  {
+    std::swap( mStartVerticalLevel, mEndVerticalLevel );
+  }
+}
+
+QgsMeshMultiLevelsFromTopAveragingMethod::QgsMeshMultiLevelsFromTopAveragingMethod()
+  : QgsMesh3dAveragingMethod( QgsMesh3dAveragingMethod::MultiLevelsFromTopAveragingMethod )
+{
+};
+
+QgsMeshMultiLevelsFromTopAveragingMethod::~QgsMeshMultiLevelsFromTopAveragingMethod() = default;
+
+QDomElement QgsMeshMultiLevelsFromTopAveragingMethod::writeXml( QDomDocument &doc ) const
+{
+  QDomElement elem = doc.createElement( QStringLiteral( "multi-vertical-layer-from-top-settings" ) );
+  elem.setAttribute( QStringLiteral( "start-layer-index" ), startVerticalLevel() );
+  elem.setAttribute( QStringLiteral( "end-layer-index" ), endVerticalLevel() );
+  return elem;
+}
+
+void QgsMeshMultiLevelsFromTopAveragingMethod::readXml( const QDomElement &elem )
+{
+  mStartVerticalLevel = elem.attribute( QStringLiteral( "start-layer-index" ) ).toInt();
+  mEndVerticalLevel = elem.attribute( QStringLiteral( "end-layer-index" ) ).toInt();
+  if ( mStartVerticalLevel > mEndVerticalLevel )
+  {
+    std::swap( mStartVerticalLevel, mEndVerticalLevel );
+  }
+}
+
+bool QgsMeshMultiLevelsFromTopAveragingMethod::equals( const QgsMesh3dAveragingMethod *other ) const
+{
+  if ( !other || other->method() != method() )
+    return false;
+
+  const QgsMeshMultiLevelsFromTopAveragingMethod *otherMethod = static_cast<const QgsMeshMultiLevelsFromTopAveragingMethod *>( other );
+
+  return ( otherMethod->startVerticalLevel() == startVerticalLevel() ) && ( otherMethod->endVerticalLevel() == endVerticalLevel() ) ;
+}
+
+QgsMesh3dAveragingMethod *QgsMeshMultiLevelsFromTopAveragingMethod::clone() const
+{
+  return new QgsMeshMultiLevelsFromTopAveragingMethod( startVerticalLevel(), endVerticalLevel() );
+}
+
+
+int QgsMeshMultiLevelsFromTopAveragingMethod::startVerticalLevel() const
+{
+  return mStartVerticalLevel;
+}
+
+int QgsMeshMultiLevelsFromTopAveragingMethod::endVerticalLevel() const
+{
+  return mEndVerticalLevel;
+}
+
+bool QgsMeshMultiLevelsFromTopAveragingMethod::hasValidInputs() const
+{
+  return mStartVerticalLevel >= 1;
```

Comment:
end vertical level probably should be also >= 1 ?

Reply:
end is always >= start (see constructor)